docs/fsharp/style-guide/component-design-guidelines.md

```diff
@@ -0,0 +1,873 @@
+
+This is useful when functions and values in the module have names that are likely to conflict with names in other modules. Requiring qualified access can greatly increase the long-term maintainability and evolvability of a library.
+
+Adding the `[<AutoOpen>]` attribute to a module means the module will be opened when the containing namespace is opened. The `[<AutoOpen>]` attribute may also be applied to an assembly to indicate a module that is automatically opened when the assembly is referenced.
+
+For example, a statistics library **MathsHeaven.Statistics** might contain a `module MathsHeaven.Statistics.Operators` containing functions `erf` and `erfc` . It is reasonable to mark this module as `[<AutoOpen>]`. This means `open MathsHeaven.Statistics` will also open this module and bring the names `erf` and `erfc` into scope. Another good use of `[<AutoOpen>]` is for modules containing extension methods.
+
+Overuse of `[<AutoOpen>]` leads to polluted namespaces, and the attribute should be used with care. For specific libraries in specific domains, judicious use of `[<AutoOpen>]` can lead to better usability.
+
+#### Consider defining operator members on classes where using well-known operators is appropriate
+
+Sometimes classes are used to model mathematical constructs such as Vectors. When the domain being modeled has well-known operators, defining them as members intrinsic to the class is helpful.
+
+```fsharp
+type Vector(x:float) =
+
+    member v.X = x
+
+    static member (*) (vector:Vector, scalar:float) = Vector(vector.X * scalar)
+
+    static member (+) (vector1:Vector, vector2:Vector) = Vector(vector1.X + vector2.X)
+
+let v = Vector(5.0)
+
+let u = v * 10.0
+```
+
+This guidance corresponds to general .NET guidance for these types. However, it can be additionally important in F# coding as this will allow these types to be used in conjunction with F# functions and methods with member constraints, such as List.sumBy.
+
+#### Consider using `[<CompiledName>]` to provide a .NET-friendly name for other .NET language consumers
+
+Sometimes you may wish name something in one style for F# consumers (such as a static member in lower case), but have a different style for the name when it is compiled into an assembly. You can use the `[<CompiledName>]` attribute to provide a different style for non F# code consuming the assembly.
+
+```fsharp
+type Vector(x:float, y:float) =
+
+    member v.X = x
+    member v.Y = y
+    
+    [<CompiledName("Create")>]
+    static member create x y = Vector (x, y)
+
+let v = Vector.create 5.0 3.0
+```
+
+By using `[<CompiledName>]`, you can use .NET naming conventions for non F# consumers of the assembly.
+
+#### Use method overloading for member functions, if doing so provides a simpler API
+
+Method overloading is a powerful tool for simplifying an API that may need to do multiple related things.
```

Comment:
In F#, it is more common to overload on number of arguments rather than types of arguments. We should probably mention this.

Reply:
Added this after the code sample, which does overload on number of arguments so as to cement this as the way to do things (generally)